File: julia/Solvers/ExhaustiveKsatSolver.jl
export ExhaustiveKsatSolver, solve

immutable ExhaustiveKsatSolver <: KsatSolver end

function solve(this:: ExhaustiveKsatSolver, problem:: KsatProblem)
  const n = problem.numVariables
  assignment = BitVector(n)
  satisfyingAssignment = uniformRandomAssignment(n)
  numSatisfyingSolutions = 0
  const numPotentialSolutions = 2^n
  for binaryAssignment:: Uint64 = 1:numPotentialSolutions
    #TODO: Could be done much faster with a Gray code or a direct conversion
    # from binary to BitVector.
    for variableIdx = 1:n
      assignment[variableIdx] = (binaryAssignment >> (variableIdx-1)) & 0x1
    end
    if checkSuccess(assignment, problem)
      if numSatisfyingSolutions == 0
        satisfyingAssignment = deepcopy(assignment)
      end
      numSatisfyingSolutions += 1
    end
  end
  return AnnotatedKsatSolution(satisfyingAssignment, numSatisfyingSolutions > 0, numSatisfyingSolutions, numPotentialSolutions)
end

File: python/ExhaustiveKsatSolver.py
import numpy as np
from KsatProblem import checkSuccess
class ExhaustiveKsatSolver():
	@staticmethod
	def solve(problem):
		n = problem.numVariables
		assignment = [False] * n
	  	satisfyingAssignment = []
	  	numSatisfyingSolutions = 0
	  	numPotentialSolutions = 2**n
	  	for binaryAssignment in range(0,numPotentialSolutions):
	  		for variableIdx in range(0,n):
	  			bit = (binaryAssignment >> (variableIdx)) & 0x1
	  			if bit == 0:
	  				assignment[variableIdx] = False
	  			else:
	  				assignment[variableIdx] = True
	  		if checkSuccess(assignment, problem):
	  			if numSatisfyingSolutions == 0:
	  				satisfyingAssignment.extend(assignment)
	  			numSatisfyingSolutions += 1
	  	return {"satisfyingAssignment":satisfyingAssignment, "isSuccessful": numSatisfyingSolutions > 0, "numSatisfyingSolutions":numSatisfyingSolutions, "numPotentialSolutions": numPotentialSolutions}

